src/main/java/build/buildfarm/instance/server/NodeInstance.java

```diff
@@ -347,22 +347,35 @@ private static boolean shouldEnsureOutputsPresent(
     // of incoming messages (perhaps due to a proxy). Or other build systems may not be reliable
     // without this extra check.
 
-    // We perform the outputs present check if the system is globally configured to check for it.
-    // Otherwise the behavior is determined dynamically from optional URI parameters.
-    if (ensureOutputsPresent) {
-      return true;
+    // The behavior is determined dynamically from optional URI parameters.
+    // If unspecified, we perform the outputs present check if the system is globally configured to
+    // check for it.
+    try {
+      URI uri = new URI(requestMetadata.getCorrelatedInvocationsId());
+      QueryStringDecoder decoder = new QueryStringDecoder(uri);
+      return decoder
+          .parameters()
+          .getOrDefault(
+              "ENSURE_OUTPUTS_PRESENT", ImmutableList.of(ensureOutputsPresent ? "true" : "false"))
+          .getFirst()
+          .equals("true");
+    } catch (URISyntaxException e) {
+      return ensureOutputsPresent;
     }
```

Comment:
Could this be extracted to a function? Looks a lot like the next block in `shouldMergeExecutions(...)`

Reply:
It also looks like the block in https://github.com/buildfarm/buildfarm/blob/main/src/main/java/build/buildfarm/server/InvocationsCollector.java#L82 for invocations. I probably just need to intercept this and put a map into the context to be extracted and sent to interested requests.